File: stack/min-stack.py
from typing import List


class MinStack:
    actualStack: List[int]
    minIndex: List[int]

    def __init__(self):
        self.actualStack = []
        self.minIndex = []

    def push(self, val: int) -> None:
        self.actualStack.append(val)
        if len(self.minIndex) == 0:
            self.minIndex.append(0)
        # New min
        elif val < self.actualStack[self.minIndex[-1]]:
            self.minIndex.append(len(self.minIndex))
        else:
            self.minIndex.append(self.minIndex[-1])

    def pop(self) -> None:
        self.actualStack.pop()
        self.minIndex.pop()

    def top(self) -> int:
        return self.actualStack[-1]

    def getMin(self) -> int:
        return self.actualStack[self.minIndex[-1]]

File: review/min-stack.cpp
#include <stack>

using namespace std;

class MinStack
{
public:
    stack<int> vals;
    stack<int> minVals;

    MinStack()
    {
        vals = stack<int>();
        minVals = stack<int>();
    }

    void push(int val)
    {
        if (vals.size() == 0)
        {
            minVals.push(val);
        }
        else
        {
            minVals.push(min(val, minVals.top()));
        }
        vals.push(val);
    }

    void pop()
    {
        vals.pop();
        minVals.pop();
    }

    int top()
    {
        return vals.top();
    }

    int getMin()
    {
        return minVals.top();
    }
};
